include/nbl/system/ISystem.h

```diff
@@ -0,0 +1,244 @@
+#ifndef __NBL_I_SYSTEM_H_INCLUDED__
+#define __NBL_I_SYSTEM_H_INCLUDED__
+
+#include <variant>
+#include "nbl/core/IReferenceCounted.h"
+#include "nbl/system/ISkippableAsyncQueueDispatcher.h"
+#include "nbl/system/IFileArchive.h"
+#include "nbl/system/IWriteFile.h"
+
+namespace nbl {
+namespace system
+{
+
+class ISystem final : public core::IReferenceCounted
+{
+public:
+    class ISystemCaller : core::IReferenceCounted
+    {
+    protected:
+        virtual ~ISystemCaller() = default;
+
+    public:
+        virtual core::smart_refctd_ptr<IReadFile> createReadFile(ISystem* sys, const std::filesystem::path& filename) = 0;
+        virtual core::smart_refctd_ptr<IWriteFile> createWriteFile(ISystem* sys, const std::filesystem::path& filename) = 0;
+        virtual uint32_t read(IReadFile* file, void* buffer, size_t offset, size_t size) = 0;
+        virtual uint32_t write(IWriteFile* file, const void* buffer, size_t offset, size_t size) = 0;
+    };
+
+protected:
+    struct SRequestType;
+    class CAsyncQueue;
+
+    class SFutureBase
+    {
+        friend ISystem;
+
+    protected:
+        bool valid_flag = false;
+        SRequestType* request;
+        CAsyncQueue* asyncQ;
+
+    public:
+        // SFuture is non-copyable and non-movable
+        SFutureBase(const SFutureBase&) = delete;
+
+        ~SFutureBase()
+        {
+            request->set_skip();
+        }
+
+        bool ready() const { return !request || request->ready; }
+        bool valid() const { return valid_flag; }
+
+        void wait()
+        {
+            asyncQ->waitForRequestCompletion(request[0]);
+        }
+    };
+
+    template <typename T>
+    class SFuture : public SFutureBase
+    {
+        alignas(T) uint8_t storage[sizeof(T)];
+
+    public:
+        ~SFuture()
+        {
+            if (valid_flag)
+                getStorage()->~T();
+        }
+
+        T* getStorage() { return reinterpret_cast<T*>(storage); }
```

Comment:
maybe make this private? or at least protected?

Reply:
yeaah, some friendship mess is needed

Reply:
This is why I suggested making a `notify()` which forwards constructor args to placement new constructor and sticking that in the `SFuture`

Reply:
resolved.